src/error.rs

```diff
@@ -1,35 +1,136 @@
 }
 
-impl From<Status> for BorrowError {
-    fn from(status: Status) -> Self {
-        BorrowError::Cairo(status)
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Error::{}",
+            match *self {
+                Error::NoMemory => "NoMemory",
+                Error::InvalidRestore => "InvalidRestore",
+                Error::InvalidPopGroup => "InvalidPopGroup",
+                Error::NoCurrentPoint => "NoCurrentPoint",
+                Error::InvalidMatrix => "InvalidMatrix",
+                Error::InvalidStatus => "InvalidStatus",
+                Error::NullPointer => "NullPointer",
+                Error::InvalidString => "InvalidString",
+                Error::InvalidPathData => "InvalidPathData",
+                Error::ReadError => "ReadError",
+                Error::WriteError => "WriteError",
+                Error::SurfaceFinished => "SurfaceFinished",
+                Error::SurfaceTypeMismatch => "SurfaceTypeMismatch",
+                Error::PatternTypeMismatch => "PatternTypeMismatch",
+                Error::InvalidContent => "InvalidContent",
+                Error::InvalidFormat => "InvalidFormat",
+                Error::InvalidVisual => "InvalidVisual",
+                Error::FileNotFound => "FileNotFound",
+                Error::InvalidDash => "InvalidDash",
+                Error::InvalidDscComment => "InvalidDscComment",
+                Error::InvalidIndex => "InvalidIndex",
+                Error::ClipNotRepresentable => "ClipNotRepresentable",
+                Error::TempFileError => "TempFileError",
+                Error::InvalidStride => "InvalidStride",
+                Error::FontTypeMismatch => "FontTypeMismatch",
+                Error::UserFontImmutable => "UserFontImmutable",
+                Error::UserFontError => "UserFontError",
+                Error::NegativeCount => "NegativeCount",
+                Error::InvalidClusters => "InvalidClusters",
+                Error::InvalidSlant => "InvalidSlant",
+                Error::InvalidWeight => "InvalidWeight",
+                Error::InvalidSize => "InvalidSize",
+                Error::UserFontNotImplemented => "UserFontNotImplemented",
+                Error::DeviceTypeMismatch => "DeviceTypeMismatch",
+                Error::DeviceError => "DeviceError",
+                Error::InvalidMeshConstruction => "InvalidMeshConstruction",
+                Error::DeviceFinished => "DeviceFinished",
+                Error::JBig2GlobalMissing => "JBig2GlobalMissing",
+                Error::PngError => "PngError",
+                Error::FreetypeError => "FreetypeError",
+                Error::Win32GdiError => "Win32GdiError",
+                Error::LastStatus => "LastStatus",
+                _ => "Unknown",
+            }
+        )
     }
 }
 
 #[derive(Error, Debug)]
 pub enum IoError {
     #[error("Cairo error: {0}")]
-    Cairo(Status),
+    Cairo(Error),
```

Comment:
```suggestion
    Cairo(#[from] Error),
```

Reply:
I'm getting 
```
error[E0658]: attributes on expressions are experimental
   --> src/error.rs:134:28
    |
134 |         BorrowError::Cairo(#[from] error)
    |                            ^^^^^^^
    |
    = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
    = help: add `#![feature(stmt_expr_attributes)]` to the crate attributes to enable

```

Reply:
How does it work one line below then? Maybe `#[from] error::Error`, maybe it needs a path...

Reply:
It only works if I remove the `From<Error> for BorrowError` & `From<Error> for IoError` implementations

Reply:
Well that makes sense, the attribute is providing exactly those `From` impls (and marking the contained error as source) :)

Reply:
Aha, makes sense now :)